algosdk/dryrun_results.py

```diff
@@ -0,0 +1,227 @@
+            lines.append(
+                [
+                    "{}".format(trace_line.pc),
+                    "{}".format(trace_line.line),
+                    truncate(src, spc.max_width),
+                    truncate(scratch, spc.max_width),
+                    truncate(stack, spc.max_width),
+                ]
+            )
+
+        return (
+            tabulate(
+                lines,
+                headers,
+                disable_numparse=True,
+                tablefmt=TableFormat(
+                    headerrow=DataRow("", " |", ""),
+                    datarow=DataRow("", " |", ""),
+                    padding=0,
+                    lineabove=None,
+                    linebelowheader=None,
+                    linebetweenrows=None,
+                    linebelow=None,
+                    with_header_hide=None,
+                ),
+            )
+            + "\n"
+        )
+
+    def app_trace(self, spc: StackPrinterConfig = None) -> str:
+        if not hasattr(self, "app_call_trace"):
+            return ""
+
+        if spc == None:
+            spc = StackPrinterConfig(top_of_stack_first=False)
+
+        return self.trace(self.app_call_trace, self.disassembly, spc=spc)
+
+    def lsig_trace(self, spc: StackPrinterConfig = None) -> str:
+        if not hasattr(self, "logic_sig_trace"):
+            return ""
+
+        if (
+            not hasattr(self, "logic_sig_disassembly")
+            or self.logic_sig_disassembly is None
+        ):
+            return ""
+
+        if spc == None:
+            spc = StackPrinterConfig(top_of_stack_first=False)
+
+        return self.trace(
+            self.logic_sig_trace, self.logic_sig_disassembly, spaces=spc
+        )
+
+
+class DryrunTrace:
+    def __init__(self, trace: List[dict]):
+        self.trace = [DryrunTraceLine(line) for line in trace]
+
+    def get_trace(self) -> List[str]:
+        return [line.trace_line() for line in self.trace]
+
+
+class DryrunTraceLine:
+    def __init__(self, tl):
+        self.line = tl["line"]
+        self.pc = tl["pc"]
+
+        self.error = ""
+        if "error" in tl:
+            self.error = tl["error"]
+
+        self.scratch = []
+        if "scratch" in tl:
+            self.scratch = [DryrunStackValue(sv) for sv in tl["scratch"]]
+
+        self.stack = [DryrunStackValue(sv) for sv in tl["stack"]]
+
+
+class DryrunStackValue:
+    def __init__(self, v):
+        self.type = v["type"]
+        self.bytes = v["bytes"]
+        self.int = v["uint"]
+
+    def __str__(self) -> str:
+        if len(self.bytes) > 0:
+            return "0x" + base64.b64decode(self.bytes).hex()
+        return str(self.int)
+
+    def __eq__(self, other: "DryrunStackValue"):
+        return (
+            self.type == other.type
+            and self.bytes == other.bytes
+            and self.int == other.int
+        )
+
+
+def truncate(s: str, max_width: int) -> str:
```

Comment:
looks like the column width is really `3 + max_width`. Is that the intention? Consistency across the API's trumps any other issues here.

Reply:
All of the SDKs have the same type of logic where we truncate at the max width, then add the 3 period ellipses. I went back and forth on this but I think having max width describe where we truncate the real value then adding ellipses to denote that the value is truncated is the right way to go. It is tough though so if you feel strongly I can change it.